crates/uv-interpreter/src/managed/find.rs

```diff
@@ -1,47 +1,140 @@
+use core::fmt;
+use fs_err as fs;
 use std::collections::BTreeSet;
 use std::ffi::OsStr;
-use std::path::PathBuf;
+use std::io;
+use std::path::{Path, PathBuf};
 use std::str::FromStr;
 
-use once_cell::sync::Lazy;
-use tracing::debug;
-
-use uv_fs::Simplified;
+use uv_state::{StateBucket, StateStore};
 
 use crate::managed::downloads::Error;
 use crate::platform::{Arch, Libc, Os};
 use crate::python_version::PythonVersion;
 
-/// The directory where Python toolchains we install are stored.
-pub static TOOLCHAIN_DIRECTORY: Lazy<Option<PathBuf>> =
-    Lazy::new(|| std::env::var_os("UV_BOOTSTRAP_DIR").map(PathBuf::from));
-
-pub fn toolchains_for_current_platform() -> Result<impl Iterator<Item = Toolchain>, Error> {
-    let platform_key = platform_key_from_env()?;
-    let iter = toolchain_directories()?
-        .into_iter()
-        // Sort "newer" versions of Python first
-        .rev()
-        .filter_map(move |path| {
-            if path
-                .file_name()
-                .map(OsStr::to_string_lossy)
-                .is_some_and(|filename| filename.ends_with(&platform_key))
-            {
-                Toolchain::new(path.clone())
-                    .inspect_err(|err| {
-                        debug!(
-                            "Ignoring invalid toolchain directory {}: {err}",
-                            path.user_display()
-                        );
+/// A collection of installed Python toolchains.
+#[derive(Debug, Clone)]
+pub struct InstalledToolchains {
+    /// The path to the top-level directory of the installed toolchains.
+    root: PathBuf,
+}
+
+impl InstalledToolchains {
+    /// A directory for installed toolchains at `root`.
+    pub fn from_path(root: impl Into<PathBuf>) -> Result<Self, io::Error> {
+        Ok(Self { root: root.into() })
+    }
+
+    /// Prefer, in order:
+    /// 1. The specific toolchain directory specified by the user, i.e., `UV_TOOLCHAIN_DIR`
+    /// 2. A bucket in the system-appropriate user-level data directory, e.g., `~/.local/uv/toolchains`
+    /// 3. A bucket in the local data directory, e.g., `./.uv/toolchains`
+    pub fn from_settings() -> Result<Self, io::Error> {
+        if let Some(toolchain_dir) = std::env::var_os("UV_TOOLCHAIN_DIR") {
+            Self::from_path(toolchain_dir)
+        } else {
+            Self::from_path(StateStore::from_settings(None)?.bucket(StateBucket::Toolchains))
+        }
+    }
+
+    pub fn temp() -> Result<Self, io::Error> {
+        Self::from_path(StateStore::temp()?.bucket(StateBucket::Toolchains))
+    }
+
+    /// Initialize the installed toolchain directory.
+    ///
+    /// Ensures the directory is created.
+    pub fn init(self) -> Result<Self, io::Error> {
+        let root = &self.root;
+
+        // Create the cache directory, if it doesn't exist.
+        fs::create_dir_all(root)?;
```

Comment:
You may want to look at the cache initialization. We probably want to add a `.gitignore` and a `CACHEDIR.tag` to this?

Reply:
I did copy this from the cache init.

I think if we add a `CACHEDIR.tag` some tools may delete this data presuming it's not important, which seems... less than ideal.

I was thinking about adding a `.gitignore` but am scared of all the bugs we encountered with it in the cache directory. I'm not sure about the `.git` boundary file and such. Hm.